microbench-ci: per metric retry

herkolategan opened the pull request:
Previously,
- The retry confirmation logic considered all metrics of a benchmark together, which could lead to false positives when different metrics regressed on different retries.
- The comparison step recalculated regression status instead of using the already-available marker files.

This PR addresses both issues by:
- Making retry logic track each metric independently, ensuring the same metric must regress consistently across retries
- Simplifying the comparison step to use marker files instead of recalculating the regression status.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: per metric retry

Previously, the retry confirmation logic took all metrics of the current
benchmark into account. This is problematic because we could possibly trigger a
regression if 3 different metrics regressed on 3 different retries for the same
benchmark. This logic has been updated to ensure the same metric has to regress
on all retries for the benchmark to be considered a regression.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: only compare the last run

Previously, if a regression occurred, the comparison step would compare all
the runs instead of just the last one. The assumption was that if all runs show
a regression we expect the combined result to show a regression, but since the
baseline case shift with each run there's a small chance this does not always
hold true. Hence, we revert here to only comparing the last run.

Epic: None
Release note: None

golgeek commented on pkg/cmd/microbench-ci/main.go:
```diff
@@ -98,6 +99,19 @@ func (s *Suite) binURL(revision Revision, benchmark *Benchmark) string {
+	artifactsDir := s.artifactsDir(revision)
+	for _, status := range []Status{Regressed, Improved} {
+		marker := strings.ToUpper(status.String())
+		markerFile := path.Join(artifactsDir, benchmark.sanitizedName()+"."+marker)
```
Nit: since you build the same file path in two different locations, you could move the code to an helper func; it would ease things if the file path is changed in the future

herkolategan replied:
Good point! I'll amend before merging.

DarrylWong reviewed (approved):
LGTM

herkolategan committed:
microbench-ci: marker file utility function

This commit adds a utility function to generate the marker file name for a
given benchmark and status.

The marker file is used to indicate that a benchmark has changed. It is
created for each metric of a benchmark.

Epic: None
Release note: None

tbg commented on pkg/cmd/microbench-ci/compare.go:
```diff
@@ -153,7 +141,10 @@ func (b Benchmarks) compareBenchmarks() (CompareResults, error) {
+
+		// If the benchmark has a marker file (regressed or improved), compare
+		// all the runs instead of just the last one.
+		if suite.hasMarkerFile(New, &benchmark) {
```
So some metric has (say) shown a regression in three consecutive runs of (say) M iterations. Here, we report the analysis on the union of the 3M runs. I _assume_ this will also be known to show a regression? In theory, the first run could've shown a regression but all the baselines may have shifted down, and in the third run they could have shifted up, relative to the middle run. The union of those may not show a regression. I know this is hypothetical, but I still wonder if it'd be cleaner to only return the last result here.

herkolategan replied:
Yeah, the thought crossed my mind. I think you're right, I'll update this to show the last run to avoid something confusing happening.

herkolategan commented on pkg/cmd/microbench-ci/testdata/regression.txt:
```diff
@@ -1,4 +1,4 @@
-# Check if summary is generated correctly
```
But, may be worthwhile to add the negative case of a partial regression 1 or 2 / 3 showing a regression that shouldn't result in a regression.

herkolategan commented:
TFTRs!

bors r=tbg,DarrylWong,golgeek
